lib/ClusterShell/Worker/Rsh.py

```diff
@@ -59,8 +61,31 @@ def _build_cmd(self):
         cmd_l.append("%s" % self.key)  # key is the node
         cmd_l.append("%s" % self.command)
 
+        # Save the return code from the command if required
+        if maxrc:
+            cmd_l.append("; echo XXRETCODE: $?")
```

Comment:
If I understand correctly, but let me know if I missed something, what you rather want is to add `"; echo XXRETCODE: $?"` to `self.command` itself. Otherwise, you will get the return code of rsh.

and this should be part of the command string, likely by doing something like the following, before appending self.command to cmd_l:

```
self.command += "; echo XXRETCODE: $?"
```

Reply:
Hello @thiell,
Perhaps I am confused.  I do not believe I really added a new Clush option.  The -S option already exists in Clush, but it did not work correctly for rsh.  All I did really was to pass that down to the Rsh Worker so that it would know when to do the "magic" since I was trying to emulate Pdsh and not do the "echo XXRETCODE: $?" if it was not required.
I did externalize the -S option as "maxrc" in the configuration file so that it would be possible to avoid passing -S on the command line on every invocation if you want that to be the default.
I am happy to submit that in multiple pull requests if you prefer. So I have a couple questions:

1) Would you prefer appending "; echo XXRETCODE: $?" for every command to the rsh worker (whether -S is passed in or not)? Happy to do it that way if you want, was just trying to avoid running the extra echo command if not requested.

2) Do you dislike having the ability to set the -S in the configuration file (using "maxrc")?  If you don't want that ability at all I am happy to remove it, leave it here, or commit it as a separate pull request.

Thanks!

Reply:
Sorry I meant a new clush option (maxrc) in the configuration file clush.conf. I understand your logic and we could indeed add `maxrc` to `clush.conf` to enable `clush -S` by default. However, I don't like the use of `task.set_info()` to pass this information to the worker. At the end, I think maxrc should stay a clush-only thing and should not impact the library/workers.

So I guess:

1. yes, definitely I would prefer that we end up with a proper rsh worker that always processes the return code. It's the same with the pdsh worker, we always gather command return codes, even if clush -S is not used. Clush then can decide if it wants to use maxrc or not, but at least we would have fixed the worker's behavior. I really thing the overhead is minimal as `echo $?` is usually a shell built-in command.

2. thanks for your flexibility! I think it's best to put the maxrc change into a different PR, so we can discuss about it separately.

If you're ok with the above, could you please update this PR so that it will focus on fixing the Rsh worker to handle command return codes, and open a separate PR to add maxrc to clush.conf that will enable `-S` by default with clush. Thanks again!

Reply:
Sure!  No problem!

Reply:
@thiell.  Thanks.  I believe that I have implemented 1.  I believe that I did what @degremont requested as well.